Migrate from Couch to SQL with bulk operations

millerdev opened the pull request:
## Product Description

- Save rows to SQL in chunks instead of saving each row individually in its own transaction.
- Do not re-process documents that have already been migrated unless --fixup-diffs is specified, and only then if there are differences.
- Legacy migrations that implement `update_or_create_sql_object` still operate the same as they did before (one document at a time).

:blowfish: Review by commit.

## Safety Assurance

### Safety story

Automated tests have been updated to verify the new bulk migration functionality. There are also existing tests for legacy migrations.

### Automated test coverage

Yes.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

millerdev committed:
Improve migration output and fix lint errors.

millerdev committed:
Add --chunk-size=N to Couch to SQL migration command

millerdev committed:
Migrate from Couch to SQL with bulk operations

- Save rows to SQL in chunks instead of saving each row individually in its own transaction.
- Do not re-process documents that have already been migrated unless --fixup-diffs is specified, and only then if there are differences.
- Legacy migrations that implement `update_or_create_sql_object` still operate the same as they did before (one document at a time).

millerdev committed:
Refactor test log file utility

millerdev committed:
Improve migration summary output

Previously it was difficult to tell if documents were missing in SQL because the totals printed were counts taken before the migration was completed.

millerdev committed:
Fix method override

millerdev commented on corehq/apps/cleanup/management/commands/populate_sql_model_from_couch_model.py:
```diff
@@ -194,11 +208,24 @@ def add_arguments(self, parser):
+            action='store_true',
+            dest='fixup_diffs',
+            default=False,
+            help="Update rows in SQL that do not match their corresponding Couch doc.",
```
Taking updates off of the default code path resolves a race condition:

- normal HQ operation loads a Couch doc
- a `PopulateSQLCommand` migration loads the same doc in a batch of 100 docs
- HQ operation modifies and saves the Couch doc, which also syncs changes to SQL (the migration's copy of the doc is now stale)
- migration calls `update_or_create_sql_object` which overwrites above changes, reverting SQL to the state of its stale Couch doc

Assuming the model sync logic is working correctly, all diffs reported by the migration should be transient, and should eventually be resolved by non-migration code if the migration only does inserts (no updates).

The `--fixup-diffs` option is provided so the migration command can be used to resolve bugs in the Couch to SQL sync logic that cause persistent differences between Couch and SQL. It is inefficient because each updated doc must be saved individually. Therefore, if a bug is discovered and most documents have already been migrated but PR 2 has not yet been merged, it may be more efficient to delete all SQL rows (since Couch is still the source of truth), fix the bug, and redo the migration.

millerdev committed:
Only print diff count when it changes

...except in legacy mode.

Processed document count is printed by with_progress_bar, so is not needed in the summary.

dannyroberts commented on corehq/apps/cleanup/management/commands/populate_sql_model_from_couch_model.py:
```diff
@@ -248,37 +277,124 @@ def handle(self, **options):
+            def iter_chunks(docs):
+                return docs  # not chunked, iterate one doc at a time
+            migrate = self._migrate_doc
+            verify = self._verify_doc
```
This took me a while to understand. I think what's confusing about it is that depending on the branch `iter_chunks`, `migrate`, and `verify` all have different signatures. `iter_chunks` for example returns either an iterable of iterables of documents or just an iterable of documents; in the later case the name also doesn't really match what it is.

This isn't show-stopping feedback, but if you can think of a way to refactor this that doesn't have confusion, great; otherwise maybe just leaving an explanation in comments under the `else:` (that unlike above functions which operate on chunks of docs, the below functions operate on single docs) would help. The inline comment in `iter_chunks` didn't quite get me there right away.

millerdev replied:
Does https://github.com/dimagi/commcare-hq/pull/32000/commits/eee6ac1ae48d001efc853a0a7a60557bde3bf503 make it easier to understand?

millerdev committed:
Fixup diffs from log file and document race condition

Helps to avoid race condition where stale Couch state is written to SQL. The race is still theoretically possible, but much less likely since document state is re-diffed before it is "fixed", and most diffs due to stale document state are transient and should be resolved before the fixup command is run.

millerdev committed:
Improve names around possibly chunked iteration

millerdev committed:
Fix ignore check on lookup table row owners migration

millerdev commented:
Added documentation about race condition. Also changed the `--fixup-diffs` option to take a log file name from which documents with diffs are retrieved, which should make it harder to hit the race condition.

millerdev commented:
Tested on staging.
